Score tests

dirkmc opened the pull request:
Part of https://github.com/libp2p/go-libp2p-pubsub/issues/287
- [x] Time in mesh
- [x] First message deliveries
- [x] Mesh message deliveries
- [x] Mesh failure penalty
- [x] Invalid message deliveries
- [x] App specific score
- [x] IP colocation factor

dirkmc committed:
test: score time in mesh

yusefnapora committed:
add TestScoreFirstMessageDeliveries

yusefnapora commented on score_test.go:
```diff
@@ -0,0 +1,134 @@
+		FirstMessageDeliveriesWeight:  1,
+		FirstMessageDeliveriesDecay: 1.0, // test without decay for now
+		FirstMessageDeliveriesCap: 2000,
+		TimeInMeshQuantum: time.Second, // bug? not setting this causes a div by zero
```
@vzyo - I hit a panic when this is unset, since we divide by it when calculating the time in mesh score.

vyzo replied:
ouch -- we should fix the validation to catch this and return an error.
once we do that let's also update the validation test to keep the score_params coverage to 100% :)

yusefnapora committed:
rm redundant initialization to zero

yusefnapora committed:
add TestScoreFirstMessageDeliveriesCap

yusefnapora committed:
add TestScoreFirstMessageDeliveriesDecay

yusefnapora committed:
fix comment

yusefnapora committed:
add TestScoreMeshMessageDeliveries

yusefnapora committed:
comments

yusefnapora committed:
add TestScoreMeshFailurePenalty

yusefnapora committed:
add TestScoreInvalidMessageDeliveries

also, TestScoreInvalidMessageDeliveriesDecay

yusefnapora committed:
add TestScoreApplicationScore

yusefnapora committed:
add TestScoreIPColocation

yusefnapora committed:
add TestScoreMeshMessageDeliveriesDecay

yusefnapora commented:
reading through, I just realized I'm using the `withinVariance` helper in places where I don't really need to... we should be able to do exact comparisons for the scores that aren't based on time.

vyzo commented:
good point.

vyzo commented:
failure in `TimeInMesh`:
```
=== RUN   TestScoreTimeInMesh

    TestScoreTimeInMesh: score_test.go:47: Score: 35.000000. Expected 20.000000 ± 10.000000

```

yusefnapora committed:
add TestScoreRetention

yusefnapora committed:
try longer interval for TimeInMesh test

we seem to take longer to collect the score on travis,
which causes the score to be outside the expected bounds.
hopefully a longer wait time will give us more wiggle room.

vyzo reviewed (approved):
looks great!
